src/par_iter/collect/consumer.rs

```diff
@@ -2,50 +2,39 @@ use super::super::len::*;
 use super::super::internal::*;
 use super::super::noop::*;
 use std::ptr;
+use std::slice;
 use std::sync::atomic::{AtomicUsize, Ordering};
 
-pub struct CollectConsumer<'c, ITEM: Send> {
+pub struct CollectConsumer<'c, ITEM: Send + 'c> {
     /// Tracks how many items we successfully wrote. Used to guarantee
     /// safety in the face of panics or buggy parallel iterators.
     writes: &'c AtomicUsize,
 
-    target: *mut ITEM,
-
-    /// Invariant: we can safely write from `target` .. `target+len`
-    len: usize,
+    /// A slice covering the target memory, not yet initialized!
+    target: &'c mut [ITEM],
```

Comment:
Hmm. I chose to use `*mut` to signal that the memory may not be initialized. However, given that there do not yet exist unsafe code guidelines that forbid this, and that I think any such guidelines would almost certainly have to accept that `&T` may point to uninitialized memory so long as you don't read from it (which would include storing to it without a `ptr::write`, it is probably ok). Then the question becomes whether indeed it is better to use `&mut[]` to track our invariants for us, but with the caveat that we must be aware of the initialization state. Probably it is.